sympy/core/tests/test_expr.py

```diff
@@ -1624,12 +1625,31 @@ def test_has_free():
     assert Integral(f(x), (f(x), 1, y)).has_free(y)
     assert not Integral(f(x), (f(x), 1, y)).has_free(x)
     assert not Integral(f(x), (f(x), 1, y)).has_free(f(x))
+    # simple extraction
+    assert (x + 1 + y).has_free(x + 1)
+    assert not (x + 2 + y).has_free(x + 1)
+    assert (2 + 3*x*y).has_free(3*x)
+    raises(ValueError, lambda: x.has_free({x, y}))
+    s = FiniteSet(1, 2)
+    assert Piecewise((s, x > 3), (4, True)).has_free(s)
+    assert not Piecewise((1, x > 3), (4, True)).has_free(s)
+    # can't make set of these, but fallback will handle
+    assert not x.has_free(y, [])
```

Comment:
Shouldn't this just give a TypeError?

Reply:
Being empty is tested before being a non-empty container.

Reply:
I'm not sure that it is yet a valid assumption that all Basic args are hashable. If it is, then an error can be raised.

Reply:
Basic args should be Basic and therefore should be hashable. If they aren't then that's a bug and it should be fixed. We shouldn't try to work around bugs like that because it just makes them less visible. There shouldn't be any code that attempts to allow for Basic to have non-Basic args.